include/pybind11/numpy.h

```diff
@@ -42,6 +43,46 @@ class array; // Forward declaration
 
 PYBIND11_NAMESPACE_BEGIN(detail)
 
+// Main author of this class: jbms@
+template <typename T>
+class LazyInitializeAtLeastOnceDestroyNever {
+public:
+    // PRECONDITION: The GIL must be held when `Get()` is called.
+    // It is possible that multiple threads execute `Get()` with `initialized_`
+    // still being false, and thus proceed to execute `initialize()`. This can
+    // happen if `initialize()` releases and reacquires the GIL internally.
+    // We accept this, and expect the operation to be both idempotent and cheap.
+    template <typename Initialize>
+    T &Get(Initialize &&initialize) {
+        if (!initialized_) {
+            assert(PyGILState_Check());
+            auto value = initialize();
+            if (!initialized_) {
+                new (reinterpret_cast<T *>(value_storage_)) T(std::move(value));
```

Comment:
The inner cast is bogus. Placement-new takes a `void*`, so you should at best cast to `void*`. However, this conversion is also available implicitly, and so unless you want to avoid hitting an overloaded placement new operator, you can just drop it:

```
new (value_storage_) T(args...);   // OK
::new (static_cast<void*>(value_storage_)) T(args...);  // Paranoidly correct
```

Reply:
This will be commit a864f2123cc83ecfc0daf755e9a70767b07de448 (see below). Sorry I missed this before triggering the currently running GHA.

I'll work on moving things to the right places before running GHA again.

```
    Semi-paranoid placement new (based on https://github.com/pybind/pybind11/pull/4877#discussion_r1350573114).

diff --git a/include/pybind11/numpy.h b/include/pybind11/numpy.h
index 47050a36..1c76177f 100644
--- a/include/pybind11/numpy.h
+++ b/include/pybind11/numpy.h
@@ -66,7 +66,7 @@ public:
             assert(PyGILState_Check());
             auto value = initialize();
             if (!initialized_) {
-                new (reinterpret_cast<T *>(value_storage_)) T(std::move(value));
+                ::new (value_storage_) T(std::move(value));
                 initialized_ = true;
             }
         }
```